File: Python/binarytreecamera.py
#Possibilites for each node to consider. 
# Possibility 1: Both children do not have the camera on them.
# Possibility 2: One child has the camera on itself
# Possibility 3: Either has no children or is not being monitored by the camera.

#We take 3 state variables: Has_Camera, Is_Monitored and Has_no_camera_or_is_not_monitored
class Solution: 
    def minCameraCover(self, root):
        camera=0
        def dfs(node):
            if not node:
                return 1
            left, right = dfs(node.left), dfs(node.right)
            if left == "Has_no_camera_or_is_not_monitored" or right == "Has_no_camera_or_is_not_monitored":
                camera +=1
                return "Has_camera"
            elif left == "Has_camera" or right == "Has_camera":
                return "Is_Monitored"
            else:
                return "Has_no_camera_or_is_not_monitored"
        if dfs(root) == "Has_no_camera_or_is_not_monitored":
            camera += 1
        return camera

File: Golang/binarytreecamera.go
//write a code for making a binary tree in go
// binary tree data structure in go

func minCameraCover(root *Treenode) int {
	camera := 0
	if minCameraCoverSub(root, &camera) == "Has_no_camera_or_camera_is_not_covered" {
		camera++
	}
	return camera
}

//Three states of a node in this binary tree
// State 0: Node has to be covered by the camera as either one of the children don't have the camera
// State 1: Node is covered by camera but does not have the camera on it
// State 2: Node is not covered by the camera and has no camera on iself

func minCameraCoverSub(root *Treenode, camera *int) string {

	if root == nil {
		return "Is_monitored"
	}
	left := minCameraCoverSub(root.Left, camera)
	right := minCameraCoverSub(root.Right, camera)

	if left == "Has_no_camera_or_is_not_monitored" || right == "Has_no_camera_or_is_not_monitored" {
		*camera++
		return "Has_camera"
	} else if left == "Has_camera" && right == "Has_camera" {
		return "Is_monitored"
	} else {
		return "Has_no_camera_or_camera_is_not_covered"
	}

}
